src/modem.rs

```diff
@@ -0,0 +1,335 @@
+use at_commands::{builder::CommandBuilder, parser::CommandParser};
+use atat::{self, AtatCmd};
+use core::{borrow::BorrowMut, ffi::c_void, marker::PhantomData};
+use esp_idf_hal::{
+    delay::{TickType, BLOCK},
+    io::EspIOError,
+    uart::{UartDriver, UartTxDriver},
+};
+
+use crate::{
+    eventloop::EspSystemEventLoop,
+    handle::RawHandle,
+    netif::{EspNetif, EspNetifDriver, NetifStack},
+    sys::*,
+};
+
+pub struct EspModem<'d, T>
+where
+    T: BorrowMut<UartDriver<'d>>,
+{
+    serial: T,
+
+    _d: PhantomData<&'d ()>,
+}
+
+impl<'d, T> EspModem<'d, T>
+where
+    T: BorrowMut<UartDriver<'d>> + Send,
+{
+    pub fn new(serial: T) -> Self {
+        Self {
+            serial,
+            _d: PhantomData,
+        }
+    }
+
+    pub fn send_cmd<CMD: AtatCmd>(&mut self, cmd: &CMD) -> Result<CMD::Response, atat::Error> {
```

Comment:
I don't think this needs to be public. Also maybe move below `setup_data_mode` which I think is/should be the main (and only) public API of your modem struct?

Reply:
I think this should be public within the `DCE` trait.